rfd/0002-custom-terraform-registry.md

```diff
@@ -0,0 +1,185 @@
+RFD.
+
+#### Why can't we just use Houston?
+
+The existing Houston distribution tool at [get.gravitational.com](https://get.gravitational.com/)
+is not suitable for distributing the files indexed by the registry. Houston has
+_very_ particular requirements about the naming of the files it serves (e.g.
+all files are expected to be `*.tar.gz` compressed tarballs - or at least
+_named_ as such) and, as it stands, will not serve the files we need it to.
+
+We _could_ modify Houston to serve the required files, but as the service is
+already slated for retirement, that does not seem to be a good use of time
+& money.
+
+### Building provider packages
+
+As far as the registry protocol is concerned, the entire package for a given
+provider consists of:
+
+* The actual binary distribution zipfile
+* A checksum file containing the SHA256 of the zipfile (formatted as per
+ `sha256sum` tool)
+* A signature file containing a binary, detached GPG signature for
+  checksum file.
+
+The Teleport provider release process currently produces a compressed tarball
+by default, and this should still be considered the primary artefact produced
+by the build, at least as long as Houston is expected to serve downloads of
+the Terraform provider.
+
+By analogy with the main Teleport release system, the registry-compatible
+provider package will be created during release promotion, with a build task
+that will:
+
+1. For each release tarball (one each for multiple release platforms):
+    1. repack the tarball into an appropriately formatted & named zipfile
+    2. generate the checksum file from the zipfile
+    3. generate the signature file by signing the checksum file
+2. Upload all generated files to the AWS bucket that backs the registry
+3. Update the registry index to include the newly released providers (see
+   _Populating the registry_, below)
+
+I expect this to be implemented as a small Go program, rather than a shell
+script. Using small Go programs for build tasks has shown itself to be a
+useful, flexible and above all _legible_ tool for scripting build tasks in
+the mainline Teleport repository, and we should continue that pattern in
+`teleport-plugins`.
+
+#### Build Security issues
+
+This process will require that the `teleport-plugins` promotion process will
+need new secrets:
+
+1. Read access to a signing key, to sign the package (see note below)
+2. Write access to the AWS bucket storing the signed packages for distribution
+
+_**Note:**_ We currently expect to use the RPM signing key as the Terraform
+signing key as it already exists and is already used for a similar purpose.
+That said, there is no technical reason why it _must_ be the RPM signing
+key. As long as the private key used to sign the package agrees with the a
+public key exposed via the registry, the signature should be deemed valid
+by Terraform.
```

Comment:
I'd like to move away from using a monolithic gpg key for signing artifiacts.  Instead, I propose we used gpg subkeys.

The setup:

1) we have a root-of-trust key signing gpg key, preferably with the private key offline, or in our highest security online storage.  This private key never touches build infra
2) We use the root-of-trust key to sign subkeys for each different artifact type (RPM, APT, TF).  These subkeys are independently exported, and used in the online build process, only for that artifact type.
3) In the even of build infra compromise, we can revoke and reissue subkeys, without having to rotatate the entire root of trust.

The sticky situation we're in currently is:

* The current root of trust isn't all that trustworthy, as it was _probably not but maybe we should treat it like it was_ compromised in https://goteleport.com/blog/hack-via-pull-request/.  There is no evidence that anyone but me attacked this, but nor can I prove the key wasn't exfiltrated via an advanced persistent threat type actor.
* Furthermore, that private key has been on every builder, and a half a dozen employees workstations/secret stores.
* Rotating keys will require customer action.  Either in a:
  * We changed keys on the official, one repo, and broke your current setup.  => "Why did things stop working?" type break.
  * We published a new repo, and the old one will stop getting updates -- better update your config.  => "Why did things fall out of date?" type break.

Anyhow, this is another topic where we probably need an RFD for the plan going forward.  Not something to change yet in this RFD, but something to be aware may change in the future.

cc @reedloden